Mathlib/NumberTheory/Harmonic/ZetaAsymp.lean

```diff
@@ -0,0 +1,373 @@
+    Tendsto (fun s ↦ riemannZeta s - 1 / (s - 1)) (𝓝[≠] 1) (𝓝 eulerMascheroniConstant) := by
+  -- We use the removable-singularity theorem to show that *some* limit over `𝓝[≠] (1 : ℂ)` exists,
+  -- and then use the previous result to deduce that this limit must be `γ`.
+  let f (s : ℂ) := riemannZeta s - 1 / (s - 1)
+  suffices ∃ C, Tendsto f (𝓝[≠] 1) (𝓝 C) by
+    cases' this with C hC
+    suffices Tendsto (fun s : ℝ ↦ f s) _ _
+      from (tendsto_nhds_unique this tendsto_riemannZeta_nhds_right) ▸ hC
+    refine hC.comp (tendsto_nhdsWithin_of_tendsto_nhds_of_eventually_within _ ?_ ?_)
+    · exact (Complex.continuous_ofReal.tendsto 1).mono_left (nhdsWithin_le_nhds ..)
+    · filter_upwards [self_mem_nhdsWithin] with a ha
+      rw [mem_compl_singleton_iff, ← Complex.ofReal_one, Ne.def, Complex.ofReal_inj]
+      exact ne_of_gt ha
+  refine ⟨_, Complex.tendsto_limUnder_of_differentiable_on_punctured_nhds_of_isLittleO ?_ ?_⟩
+  · filter_upwards [self_mem_nhdsWithin] with s hs
+    refine (differentiableAt_riemannZeta hs).sub ((differentiableAt_const _).div ?_ ?_)
+    · fun_prop
+    · rwa [mem_compl_singleton_iff, ← sub_ne_zero] at hs
+  · refine Asymptotics.isLittleO_of_tendsto' ?_ ?_
+    · filter_upwards [self_mem_nhdsWithin] with t ht ht'
+      rw [inv_eq_zero, sub_eq_zero] at ht'
+      tauto
+    · simp_rw [div_eq_mul_inv, inv_inv, sub_mul,
+        (by ring_nf : 𝓝 (0 : ℂ) = 𝓝 ((1 - 1) - f 1 * (1 - 1)))]
+      apply Tendsto.sub
+      · simp_rw [mul_comm (f _), f, mul_sub]
+        apply riemannZeta_residue_one.sub
+        refine Tendsto.congr' ?_ (tendsto_const_nhds.mono_left nhdsWithin_le_nhds)
+        filter_upwards [self_mem_nhdsWithin] with x hx
+        field_simp [sub_ne_zero.mpr <| mem_compl_singleton_iff.mp hx]
+      · exact ((tendsto_id.sub tendsto_const_nhds).mono_left nhdsWithin_le_nhds).const_mul _
+
+lemma _root_.isBigO_nhds_ne_iff {α E F : Type*} [TopologicalSpace α] [Norm E] [SeminormedAddGroup F]
+    {f : α → E} {g : α → F} {a : α} (h : ‖g a‖ = 0 → ‖f a‖ = 0) :
+    f =O[𝓝[≠] a] g ↔ f =O[𝓝 a] g := by
+  refine ⟨fun h ↦ ?_, fun h ↦ h.mono nhdsWithin_le_nhds⟩
+  simp only [Asymptotics.isBigO_iff, eventually_nhdsWithin_iff] at h ⊢
+  obtain ⟨c, hc⟩ := h
+  use max c (‖f a‖ / ‖g a‖)
+  filter_upwards [hc] with b hb
+  rcases eq_or_ne a b with rfl | hb'
+  · transitivity (‖f a‖ / ‖g a‖) * ‖g a‖
+    · by_cases hf : ‖f a‖ = 0
+      · rw [hf, zero_div, zero_mul]
+      · exact le_of_eq (div_mul_cancel₀ _ (show ‖g a‖ ≠ 0 by tauto)).symm
+    · simp only [max_mul_of_nonneg _ _ (norm_nonneg _), le_max_iff, le_refl, or_true]
+  · refine (hb hb'.symm).trans ?_
+    simp only [max_mul_of_nonneg _ _ (norm_nonneg _), le_max_iff, le_refl, true_or]
+
+lemma _root_.isBigO_one_nhds_ne_iff {α E F : Type*} [TopologicalSpace α]
+    [Norm E] [SeminormedAddCommGroup F] [One F] [NormOneClass F]
+    {f : α → E} (a : α) :
+    f =O[𝓝[≠] a] (fun _ ↦ 1 : α → F) ↔ f =O[𝓝 a] (fun _ ↦ 1 : α → F) :=
+  isBigO_nhds_ne_iff (by simp)
```

Comment:
Can you move these two lemmas elsewhere (probably a file on asymptotics)? They have nothing specific to the Riemann zeta function.